database.c

```diff
@@ -480,3 +481,164 @@ void *DB_thread(void *val)
 
 	return NULL;
 }
+
+// Get most recent 24 hours data from long-term database
+void read_data_from_DB(void)
+{
+	// Open database file
+	if(!dbopen())
+	{
+		logg("read_data_from_DB() - Failed to open DB");
+		return;
+	}
+
+	// Prepare request
+	char *rstr = NULL;
+	// Get time stamp 24 hours in the past
+	int differencetofullhour = time(NULL) % GCinterval;
+	long int mintime = ((long)time(NULL) - GCdelay - differencetofullhour) - MAXLOGAGE;
+	int rc = asprintf(&rstr, "SELECT * FROM queries WHERE timestamp >= %li", mintime);
+	if(rc < 1)
+	{
+		logg("read_data_from_DB() - Allocation error (%i): %s", rc, sqlite3_errmsg(db));
+		return;
+	}
+	// Log DB query string in debug mode
+	if(debug) logg(rstr);
+
+	// Prepare SQLite3 statement
+	sqlite3_stmt* stmt;
+	rc = sqlite3_prepare_v2(db, rstr, -1, &stmt, NULL);
+	if( rc ){
+		logg("read_data_from_DB() - SQL error prepare (%i): %s", rc, sqlite3_errmsg(db));
+		dbclose();
+		check_database(rc);
+		return;
+	}
+
+	// Loop through returned database rows
+	while((rc = sqlite3_step(stmt)) == SQLITE_ROW)
+	{
+		// Ensure we have enough space in the queries struct
+		memory_check(QUERIES);
+		memory_check(DOMAINS);
+		memory_check(CLIENTS);
+
+		// Set ID for this query
+		int queryID = counters.queries;
+
+		int queryTimeStamp = sqlite3_column_int(stmt, 1);
+		int type = sqlite3_column_int(stmt, 2);
+		int status = sqlite3_column_int(stmt, 3);
+		int domainID = findDomainID((const char *)sqlite3_column_text(stmt, 4));
+		int clientID = findClientID((const char *)sqlite3_column_text(stmt, 5));
+		int forwardID = findForwardID((const char *)sqlite3_column_text(stmt, 6), true);
+
+		int overTimeTimeStamp = queryTimeStamp - (queryTimeStamp % 600 + 300);
+		int timeidx = findOverTimeID(overTimeTimeStamp);
+		validate_access("overTime", timeidx, true, __LINE__, __FUNCTION__, __FILE__);
+
+		// Store this query in memory
+		validate_access("queries", queryID, false, __LINE__, __FUNCTION__, __FILE__);
+		queries[queryID].magic = MAGICBYTE;
+		queries[queryID].timestamp = queryTimeStamp;
+		queries[queryID].type = type;
+		queries[queryID].status = status;
+		queries[queryID].domainID = domainID;
+		queries[queryID].clientID = clientID;
+		queries[queryID].forwardID = forwardID;
+		queries[queryID].timeidx = timeidx;
+		queries[queryID].valid = true; // Mark this as a valid query (false = it has been garbage collected and should be skipped)
+		queries[queryID].db = true; // Mark this as already present in the database
+		queries[queryID].id = 0; // This is dnsmasq's internal ID. We don't store it in the database
```

Comment:
Just to make sure, the dnsmasq query IDs don't start at 0, right?

Reply:
Yes, it starts with `1`.